src/main/java/com/scalar/cassy/transferer/AzureBlobFileDownloader.java

```diff
@@ -27,33 +29,33 @@ public AzureBlobFileDownloader(BlobContainerAsyncClient blobContainerClient) {
   @Override
   public void download(RestoreConfig config) {
     String key = BackupPath.create(config, config.getKeyspace());
-    List<CompletableFuture> futures = new ArrayList<>();
-    blobContainerClient
-        .listBlobs(new ListBlobsOptions().setPrefix(key))
-        .toIterable()
-        .forEach(
-            b -> {
-              try {
-                Files.createDirectories(Paths.get(config.getDataDir(), b.getName()).getParent());
-              } catch (IOException e) {
-                throw new FileTransferException(e);
-              }
-              futures.add(
-                  blobContainerClient
-                      .getBlobAsyncClient(b.getName())
-                      .downloadToFile(Paths.get(config.getDataDir(), b.getName()).toString())
-                      .toFuture());
-            });
 
-    logger.info("Downloading " + blobContainerClient.getBlobContainerName() + key);
-    futures.forEach(
-        f -> {
-          try {
-            f.get();
-          } catch (InterruptedException | ExecutionException e) {
-            throw new FileTransferException(e);
-          }
-        });
+    logger.info("Downloading files of '" + config.getKeyspace() + "' keyspace");
+    List<Mono<BlobProperties>> filesToBeDownloaded = new ArrayList<>();
+    Iterable<BlobItem> keyspaceBlobs =
+        blobContainerClient.listBlobs(new ListBlobsOptions().setPrefix(key + "/")).toIterable();
+    for (BlobItem blob : keyspaceBlobs) {
+      Path destFile = Paths.get(config.getDataDir(), blob.getName());
+      try {
+        Files.createDirectories(destFile.getParent());
+      } catch (IOException e) {
+        throw new FileTransferException(e);
+      }
+      filesToBeDownloaded.add(
+          blobContainerClient
+              .getBlobAsyncClient(blob.getName())
+              .downloadToFile(destFile.toString())
+              .doOnSuccess(
+                  blobProperties -> logger.info("Download file succeeded : " + destFile.toString()))
+              .doOnError(error -> logger.error("Download failed : " + destFile.toString(), error)));
```

Comment:
Just a question. So if it fails, it triggers `doOnError` and the files won't be added to `filesToBeDownloaded` ?
Not sure what would happen in case of failure.
`Mono.when` will throw an exception ?

Reply:
In this block, the `Mono` is only created then added to the `filesToBeDownloaded` list. The `doOnSuccess` and `doOnError` callback will be executed only when the Mono execution has been processed.
When the mono is processed and in the case it throws an exception, the `doOnError` will be called but the exception will still be propagated. Besides, I can configured the Mono to not propagate the exception if you would like.
The call to `Mono.when(filesToBeDownloaded).block()` will process each mono of the list. If one of them fail, the program will terminate immediately which would stop any still other ongoing mono. 

Your question made me realize that the current code in case of an error will end up printing twice the error in the log. 
The first will be printed by the `doOnError` callback and the second by the `FileTransferException(e);` bubbling up the stack. 
I propose to update the code to simply throw a `FileTransferException` in the `doOnError` instead and remove the 
```
} catch (Exception e) {
      throw new FileTransferException(e);
    }
``` 
around the `Mono.when(filesToBeDownloaded).block();`